File: web-app/src/features/instrumentation/api/instrumentation-v2-contract.test.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0.
 */

import { describe, expect, it } from 'vitest';

import {
  parseCatalogResponse,
  parseDetectionResponse,
  parseIntakeProfilesResponse,
  parseRenderResponse
} from './instrumentation-v2-wire';

const component = {
  name: 'OpenTelemetry',
  sourceUrl: 'https://github.com/open-telemetry/opentelemetry-collector-contrib',
  version: '0.130.0',
  versionPolicy: 'pinned',
  license: 'Apache-2.0',
  installationLocationKey: 'instrumentation.location.application',
  official: true,
  bundledWithHertzBeat: false,
  dependencies: [],
  artifacts: []
};

const service = { name: 'checkout', namespace: 'shop', environment: 'production' };
const jumpContext = {
  serviceName: 'checkout',
  serviceNamespace: 'shop',
  environment: 'production',
  intakeProfileId: 'server-default',
  startedAt: 1000,
  detectedAt: 2000
};

describe('instrumentation v2 wire contracts', () => {
  it('parses the backend-owned source directory without accepting legacy source shapes', () => {
    const value = parseCatalogResponse({
      schemaVersion: 2,
      groups: [
        { id: 'quick_start', labelKey: 'instrumentation.v2.directory.group.quick_start' },
        { id: 'applications', labelKey: 'instrumentation.v2.directory.group.applications' }
      ],
      sources: [
        {
          id: 'quick_start',
          labelKey: 'instrumentation.v2.directory.source.quick_start',
          descriptionKey: 'instrumentation.v2.directory.source.quick_start_description',
          iconKey: 'quick-start',
          groupIds: ['quick_start'],
          support: 'supported',
          sourceKind: 'quick_start',
          recipeIds: ['opentelemetry_telemetrygen'],
          signals: { metrics: 'supported', logs: 'supported', traces: 'supported' },
          documentationUrl: 'https://opentelemetry.io/docs/'
        },
        {
          id: 'java',
          labelKey: 'instrumentation.v2.directory.source.java',
          descriptionKey: 'instrumentation.v2.directory.source.java_description',
          iconKey: 'java',
          groupIds: ['applications'],
          support: 'supported',
          sourceKind: 'application',
          recipeIds: ['java_spring'],
          signals: { metrics: 'supported', logs: 'preview', traces: 'supported' },
          documentationUrl: 'https://opentelemetry.io/docs/languages/java/'
        }
      ],
      recipes: [
        {
          id: 'opentelemetry_telemetrygen',
          kind: 'quick_start',
          labelKey: 'instrumentation.recipe.telemetrygen',
          preview: false,
          environments: ['docker'],
          platforms: ['linux_amd64'],
          signals: { metrics: 'supported', logs: 'supported', traces: 'supported' },
          components: [component],
          blocksPreview: ['command', 'environment', 'check']
        },
        {
          id: 'java_spring',
          kind: 'application',
          labelKey: 'instrumentation.recipe.java_spring',
          preview: false,
          language: 'java',
          framework: 'spring_boot',
          method: 'zero_code',
          environments: ['docker'],
          platforms: ['linux_amd64'],
          signals: { metrics: 'supported', logs: 'preview', traces: 'supported' },
          components: [component],
          blocksPreview: ['environment']
        }
      ]
    });
    expect(Object.keys(value)).toEqual(['schemaVersion', 'groups', 'sources', 'recipes']);
    expect(value.sources[1]?.recipeIds).toEqual(['java_spring']);
    expect(value.recipes[0]?.id).toBe('opentelemetry_telemetrygen');
  });

  it('keeps unsupported catalog entries discoverable but non-selectable', () => {
    const source = parseCatalogResponse({
      schemaVersion: 2,
      groups: [{ id: 'logs', labelKey: 'instrumentation.v2.directory.group.logs' }],
      sources: [
        {
          id: 'fluent_bit',
          labelKey: 'instrumentation.v2.directory.source.fluent_bit',
          descriptionKey: 'instrumentation.v2.directory.source.fluent_bit_description',
          iconKey: 'fluent-bit',
          groupIds: ['logs'],
          support: 'unsupported',
          recipeIds: [],
          signals: { metrics: 'unsupported', logs: 'unsupported', traces: 'unsupported' }
        }
      ],
      recipes: []
    }).sources[0];
    expect(source).toMatchObject({ id: 'fluent_bit', support: 'unsupported' });
    expect(source).not.toHaveProperty('sourceKind');
  });

  it('rejects unknown source groups and inconsistent unsupported entries', () => {
    const base = {
      schemaVersion: 2,
      groups: [{ id: 'logs', labelKey: 'instrumentation.v2.directory.group.logs' }],
      recipes: []
    };
    const source = {
      id: 'fluent_bit',
      labelKey: 'instrumentation.v2.directory.source.fluent_bit',
      descriptionKey: 'instrumentation.v2.directory.source.fluent_bit_description',
      iconKey: 'fluent-bit',
      groupIds: ['missing'],
      support: 'unsupported',
      recipeIds: [],
      signals: { metrics: 'unsupported', logs: 'unsupported', traces: 'unsupported' }
    };
    expect(() => parseCatalogResponse({ ...base, sources: [source] })).toThrow();
    expect(() =>
      parseCatalogResponse({
        ...base,
        sources: [{ ...source, groupIds: ['logs'], sourceKind: 'application', recipeIds: ['java_spring'] }]
      })
    ).toThrow();
    expect(() =>
      parseCatalogResponse({
        ...base,
        sources: [
          {
            ...source,
            groupIds: ['logs'],
            support: 'preview',
            sourceKind: 'existing_opentelemetry',
            recipeIds: ['missing']
          }
        ]
      })
    ).toThrow();
  });

  it.each([
    { schemaVersion: 2, status: 'unconfigured', profiles: [] },
    {
      schemaVersion: 2,
      status: 'unavailable',
      errorCode: 'intake_profile_discovery_unavailable',
      profiles: []
    }
  ])('preserves the frozen profile discovery state %#', value => {
    expect(parseIntakeProfilesResponse(value)).toEqual(value);
  });

  it('rejects unavailable discovery disguised as an empty available result', () => {
    expect(() => parseIntakeProfilesResponse({ schemaVersion: 2, status: 'available', profiles: [] })).toThrow();
  });

  it.each([
    ['https://example.test/otlp', 'plaintext'],
    ['http://example.test/otlp', 'tls'],
    ['ftp://example.test/otlp', 'plaintext'],
    ['http://', 'plaintext'],
    ['http:example.test/otlp', 'plaintext'],
    ['http:/example.test/otlp', 'plaintext'],
    ['http:///example.test/otlp', 'plaintext'],
    [' http://example.test/otlp', 'plaintext'],
    ['https://user@example.test/otlp', 'tls'],
    ['https://example.test/otlp?token=secret', 'tls'],
    ['https://example.test/otlp#fragment', 'tls']
  ])('rejects endpoint/security mismatch or unsafe URL %s', (url, security) => {
    expect(() =>
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        defaultProfileId: 'server-default',
        profiles: [
          {
            id: 'server-default',
            kind: 'server',
            availability: 'available',
            gateway: 'server',
            supportedTransports: ['http_protobuf'],
            endpoints: { http_protobuf: { url, security } },
            authentication: 'bearer_token',
            authorizationHeader: 'Authorization'
          }
        ]
      })
    ).toThrow();
  });

  it.each([
    ['https://example.test/otlp', 'tls'],
    ['http://example.test/otlp', 'plaintext']
  ])('preserves backend endpoint security evidence for %s', (url, security) => {
    expect(
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        defaultProfileId: 'server-default',
        profiles: [
          {
            id: 'server-default',
            kind: 'server',
            availability: 'available',
            gateway: 'server',
            supportedTransports: ['http_protobuf'],
            endpoints: { http_protobuf: { url, security } },
            authentication: 'bearer_token',
            authorizationHeader: 'Authorization'
          }
        ]
      }).profiles[0]?.endpoints.http_protobuf
    ).toEqual({ url, security });
  });

  it('rejects the removed httpsEndpoints field and unavailable advertised endpoints', () => {
    expect(() =>
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        profiles: [
          {
            id: 'server-default',
            kind: 'server',
            availability: 'available',
            gateway: 'server',
            supportedTransports: ['http_protobuf'],
            httpsEndpoints: { http_protobuf: 'https://example.test/otlp' },
            authentication: 'bearer_token',
            authorizationHeader: 'Authorization'
          }
        ]
      })
    ).toThrow();
    expect(() =>
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        profiles: [
          {
            id: 'collector-edge',
            kind: 'hertzbeat_collector',
            availability: 'unavailable',
            supportedTransports: [],
            endpoints: {
              http_protobuf: { url: 'http://collector.test:4318', security: 'plaintext' }
            },
            collectorId: 'edge',
            errorCode: 'intake_profile_unavailable'
          }
        ]
      })
    ).toThrow();
  });

  it('rejects profile kind/gateway and Collector identity mismatches', () => {
    expect(() =>
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        profiles: [
          {
            id: 'collector-edge',
            kind: 'hertzbeat_collector',
            availability: 'available',
            gateway: 'server',
            supportedTransports: ['grpc'],
            endpoints: { grpc: { url: 'https://example.test:4317', security: 'tls' } },
            authentication: 'bearer_token',
            authorizationHeader: 'Authorization'
          }
        ]
      })
    ).toThrow();
  });

  it('allows an unavailable Collector identity without advertised connectivity', () => {
    expect(
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        profiles: [
          {
            id: 'collector-edge',
            kind: 'hertzbeat_collector',
            availability: 'unavailable',
            supportedTransports: [],
            endpoints: {},
            collectorId: 'edge',
            errorCode: 'intake_profile_unavailable'
          }
        ]
      }).profiles[0]?.collectorId
    ).toBe('edge');
  });

  it('parses an explicitly unauthenticated external destination without legacy authorization metadata', () => {
    const response = parseIntakeProfilesResponse({
      schemaVersion: 2,
      status: 'available',
      defaultProfileId: 'external-local',
      profiles: [
        {
          id: 'external-local',
          kind: 'external_otel_collector',
          availability: 'available',
          gateway: 'external',
          supportedTransports: ['http_protobuf'],
          endpoints: { http_protobuf: { url: 'http://otel.example.test:4318', security: 'plaintext' } },
          authentication: 'none'
        }
      ]
    });

    expect(response.profiles[0]).toMatchObject({
      authentication: 'none',
      authorizationHeader: null
    });
    expect(JSON.stringify(response)).not.toContain('authHeaderName');
  });

  it.each([
    { authentication: 'none', authorizationHeader: 'Authorization' },
    { authentication: 'bearer_token', authorizationHeader: null },
    { authentication: 'bearer_token', authorizationHeader: 'X-Authorization' },
    { authentication: 'none', authHeaderName: 'Authorization' }
  ])('rejects inconsistent or legacy authentication metadata %#', authentication => {
    expect(() =>
      parseIntakeProfilesResponse({
        schemaVersion: 2,
        status: 'available',
        defaultProfileId: 'external-local',
        profiles: [
          {
            id: 'external-local',
            kind: 'external_otel_collector',
            availability: 'available',
            gateway: 'external',
            supportedTransports: ['http_protobuf'],
            endpoints: { http_protobuf: { url: 'http://otel.example.test:4318', security: 'plaintext' } },
            ...authentication
          }
        ]
      })
    ).toThrow();
  });

  it.each([
    'http://example.test/docs',
    'https://user@example.test/docs',
    'https://example.test/docs?token=secret',
    'https://example.test/docs#fragment'
  ])('rejects unsafe guide link %s', href => {
    expect(() =>
      parseRenderResponse({
        schemaVersion: 2,
        sourceKind: 'existing_opentelemetry',
        recipeId: 'existing_otlp',
        intakeProfile: {
          id: 'server-default',
          kind: 'server',
          availability: 'available',
          gateway: 'server',
          supportedTransports: ['http_protobuf'],
          endpoints: { http_protobuf: { url: 'https://example.test/otlp', security: 'tls' } },
          authentication: 'bearer_token',
          authorizationHeader: 'Authorization'
        },
        service,
        signals: { metrics: 'supported', logs: 'supported', traces: 'supported' },
        components: [],
        secretPlaceholders: {
          authorizationToken: { marker: '${HERTZBEAT_TOKEN}', kind: 'authorization_token' }
        },
        blocks: [
          {
            id: 'docs',
            type: 'link',
            titleKey: 'instrumentation.v2.block.docs',
            executionLocationKey: 'instrumentation.location.hertzbeat_ui',
            href,
            placeholders: []
          }
        ]
      })
    ).toThrow();
  });

  it('parses generic guide blocks without materializing the token', () => {
    const value = parseRenderResponse({
      schemaVersion: 2,
      sourceKind: 'quick_start',
      recipeId: 'opentelemetry_telemetrygen',
      intakeProfile: {
        id: 'server-default',
        kind: 'server',
        availability: 'available',
        gateway: 'server',
        supportedTransports: ['http_protobuf'],
        endpoints: { http_protobuf: { url: 'https://example.test/v1/otlp', security: 'tls' } },
        authentication: 'bearer_token',
        authorizationHeader: 'Authorization'
      },
      service,
      signals: { metrics: 'supported', logs: 'supported', traces: 'supported' },
      components: [component],
      secretPlaceholders: {
        authorizationToken: { marker: '${HERTZBEAT_TOKEN}', kind: 'authorization_token' }
      },
      blocks: [
        {
          id: 'plaintext_transport_warning',
          type: 'warning',
          titleKey: 'instrumentation.v2.block.plaintext_transport_warning',
          bodyKey: 'instrumentation.v2.warning.plaintext_authorization',
          executionLocationKey: 'instrumentation.location.hertzbeat',
          placeholders: []
        },
        {
          id: 'send',
          type: 'command',
          titleKey: 'instrumentation.block.send',
          executionLocationKey: 'instrumentation.location.terminal',
          language: 'shell',
          content: 'Authorization=${HERTZBEAT_TOKEN}',
          placeholders: ['authorizationToken']
        },
        {
          id: 'verify',
          type: 'check',
          titleKey: 'instrumentation.block.verify',
          bodyKey: 'instrumentation.v2.block.verify_body',
          executionLocationKey: 'instrumentation.location.hertzbeat',
          placeholders: []
        }
      ]
    });
    expect(value.blocks[0]).toMatchObject({ id: 'plaintext_transport_warning', type: 'warning' });
    expect(value.blocks[1]?.content).toContain('${HERTZBEAT_TOKEN}');
    expect(JSON.stringify(value)).not.toContain('secret-value');
  });

  it('accepts a no-auth guide only when token declarations, markers, and Authorization content are absent', () => {
    const value = parseRenderResponse({
      schemaVersion: 2,
      sourceKind: 'existing_opentelemetry',
      recipeId: 'existing_otlp',
      intakeProfile: {
        id: 'external-local',
        kind: 'external_otel_collector',
        availability: 'available',
        gateway: 'external',
        supportedTransports: ['http_protobuf'],
        endpoints: { http_protobuf: { url: 'http://otel.example.test:4318', security: 'plaintext' } },
        authentication: 'none'
      },
      service,
      signals: { metrics: 'supported', logs: 'supported', traces: 'supported' },
      components: [],
      secretPlaceholders: {},
      blocks: [
        {
          id: 'configure_exporter',
          type: 'code',
          titleKey: 'instrumentation.v2.block.configure_exporter',
          executionLocationKey: 'instrumentation.location.otel_collector',
          language: 'yaml',
          content: 'endpoint: http://otel.example.test:4318',
          placeholders: []
        }
      ]
    });

    expect(value.intakeProfile).toMatchObject({ authentication: 'none', authorizationHeader: null });
    expect(value.secretPlaceholders).toEqual({});
    expect(value.blocks[0]?.placeholders).toEqual([]);

    expect(() =>
      parseRenderResponse({
        ...value,
        blocks: [{ ...value.blocks[0]!, content: 'Authorization: Bearer ${HERTZBEAT_TOKEN}' }]
      })
    ).toThrow();
    expect(() =>
      parseRenderResponse({
        ...value,
        intakeProfile: {
          ...value.intakeProfile,
          availability: 'unavailable',
          gateway: undefined,
          supportedTransports: [],
          endpoints: {},
          authentication: undefined,
          errorCode: 'intake_profile_unavailable'
        },
        blocks: []
      })
    ).toThrow();
  });

  it('requires exactly three backend query jumps and preserves their contexts', () => {
    const response = parseDetectionResponse({
      schemaVersion: 2,
      detectedAt: 2000,
      context: {
        sourceKind: 'quick_start',
        recipeId: 'opentelemetry_telemetrygen',
        service,
        intakeProfileId: 'server-default',
        startedAt: 1000,
        windowEndAt: 121000
      },
      signals: {
        metrics: { status: 'received', lastReceivedAt: 1900 },
        logs: { status: 'waiting', errorCode: 'signal_not_received' },
        traces: { status: 'unsupported', errorCode: 'signal_not_supported' }
      },
      polling: { decision: 'continue_polling', pollAfterMs: 3000, deadlineAt: 121000 },
      queryJumpContext: jumpContext,
      queryJumps: [
        { signal: 'metrics', enabled: true, context: jumpContext },
        { signal: 'logs', enabled: false, context: jumpContext },
        { signal: 'traces', enabled: false, context: jumpContext }
      ]
    });
    expect(response.queryJumps.map(jump => jump.signal)).toEqual(['metrics', 'logs', 'traces']);
    expect(() => parseDetectionResponse({ ...response, queryJumps: response.queryJumps.slice(0, 2) })).toThrow();
    expect(() =>
      parseDetectionResponse({
        ...response,
        queryJumps: response.queryJumps.map((jump, index) => (index === 0 ? { ...jump, enabled: false } : jump))
      })
    ).toThrow();
    expect(() =>
      parseDetectionResponse({
        ...response,
        queryJumps: response.queryJumps.map((jump, index) =>
          index === 0 ? { ...jump, context: { ...jump.context, serviceName: 'other' } } : jump
        )
      })
    ).toThrow();
  });

  it('rejects detection evidence that escapes the safe scoped status contract', () => {
    const response = detectionResponse();
    expect(() =>
      parseDetectionResponse({
        ...response,
        signals: {
          ...response.signals,
          metrics: { status: 'received', lastReceivedAt: response.detectedAt + 1 }
        }
      })
    ).toThrow();
    expect(() =>
      parseDetectionResponse({
        ...response,
        signals: {
          ...response.signals,
          logs: { status: 'unavailable', errorCode: 'database_password_leaked' }
        }
      })
    ).toThrow();
    expect(() =>
      parseDetectionResponse({
        ...response,
        polling: { decision: 'complete', deadlineAt: response.context.windowEndAt }
      })
    ).toThrow();
  });

  it('binds query jumps to the complete echoed detection context', () => {
    const response = detectionResponse();
    const mismatchedJumpContext = {
      ...response.queryJumpContext,
      collectorId: 'other-collector',
      serviceInstanceId: 'other-instance',
      endpoint: '/other'
    };
    expect(() =>
      parseDetectionResponse({
        ...response,
        queryJumpContext: mismatchedJumpContext,
        queryJumps: response.queryJumps.map(jump => ({ ...jump, context: mismatchedJumpContext }))
      })
    ).toThrow();
  });
});

function detectionResponse() {
  const scopedService = {
    ...service,
    serviceInstanceId: 'checkout-1',
    endpoint: '/checkout/{id}'
  };
  const scopedJump = {
    serviceName: scopedService.name,
    serviceNamespace: scopedService.namespace,
    environment: scopedService.environment,
    intakeProfileId: 'collector-edge',
    collectorId: 'edge',
    serviceInstanceId: scopedService.serviceInstanceId,
    endpoint: scopedService.endpoint,
    startedAt: 1000,
    detectedAt: 2000
  };
  return {
    schemaVersion: 2 as const,
    detectedAt: 2000,
    context: {
      sourceKind: 'quick_start' as const,
      recipeId: 'opentelemetry_telemetrygen',
      service: scopedService,
      intakeProfileId: 'collector-edge',
      collectorId: 'edge',
      startedAt: 1000,
      windowEndAt: 121000
    },
    signals: {
      metrics: { status: 'received' as const, lastReceivedAt: 1900 },
      logs: { status: 'waiting' as const, errorCode: 'signal_not_received' },
      traces: { status: 'unsupported' as const, errorCode: 'signal_not_supported' }
    },
    polling: { decision: 'continue_polling' as const, pollAfterMs: 3000, deadlineAt: 121000 },
    queryJumpContext: scopedJump,
    queryJumps: [
      { signal: 'metrics' as const, enabled: true, context: scopedJump },
      { signal: 'logs' as const, enabled: false, context: scopedJump },
      { signal: 'traces' as const, enabled: false, context: scopedJump }
    ]
  };
}

File: hertzbeat-observability/src/test/java/org/apache/hertzbeat/observability/instrumentation/v2/InstrumentationV2ContractTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.hertzbeat.observability.instrumentation.v2;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URI;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.apache.hertzbeat.observability.instrumentation.api.InstrumentationApiContract.DetectionErrorCode;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationCatalogV2.SourceKind;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationDetectionV2.DetectionStatus;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationDetectionV2.PollingDecision;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationDetectionV2.PollingInstruction;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationDetectionV2.SignalDetection;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationGuideV2.BlockType;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationGuideV2.SecretPlaceholder;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.Availability;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.Authentication;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.Gateway;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.IntakeEndpoint;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.IntakeKind;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.IntakeProfile;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.OtlpTransport;
import org.apache.hertzbeat.observability.instrumentation.v2.api.InstrumentationIntakeProfileV2.TransportSecurity;
import org.apache.hertzbeat.observability.instrumentation.v2.controller.InstrumentationV2Controller;
import org.apache.hertzbeat.observability.instrumentation.service.InstrumentationCatalogService;
import org.apache.hertzbeat.observability.instrumentation.v2.service.InstrumentationCatalogV2Service;
import org.junit.jupiter.api.Test;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;

class InstrumentationV2ContractTest {

    private final ObjectMapper mapper = new ObjectMapper();

    @Test
    void freezesCanonicalSurfaceAndClosedWireEnums() throws Exception {
        RequestMapping mapping = InstrumentationV2Controller.class.getAnnotation(RequestMapping.class);
        assertEquals(List.of("/api/instrumentation"), List.of(mapping.path()));
        assertEquals(
                Set.of("/catalog", "/intake-profiles"),
                java.util.Arrays.stream(InstrumentationV2Controller.class.getDeclaredMethods())
                        .map(method -> method.getAnnotation(GetMapping.class))
                        .filter(java.util.Objects::nonNull)
                        .flatMap(annotation -> java.util.Arrays.stream(annotation.value()))
                        .collect(java.util.stream.Collectors.toUnmodifiableSet()));
        assertEquals(
                Set.of("/render", "/detect"),
                java.util.Arrays.stream(InstrumentationV2Controller.class.getDeclaredMethods())
                        .map(method -> method.getAnnotation(PostMapping.class))
                        .filter(java.util.Objects::nonNull)
                        .flatMap(annotation -> java.util.stream.Stream.concat(
                                java.util.Arrays.stream(annotation.value()),
                                java.util.Arrays.stream(annotation.path())))
                        .collect(java.util.stream.Collectors.toUnmodifiableSet()));
        assertEquals("\"quick_start\"", mapper.writeValueAsString(SourceKind.QUICK_START));
        assertEquals("\"application\"", mapper.writeValueAsString(SourceKind.APPLICATION));
        assertEquals("\"existing_opentelemetry\"", mapper.writeValueAsString(SourceKind.EXISTING_OPENTELEMETRY));
        assertEquals(
                List.of("command", "code", "environment", "download", "note", "warning", "link", "check"),
                java.util.Arrays.stream(BlockType.values())
                        .map(value -> mapper.convertValue(value, String.class))
                        .toList());
        assertEquals(
                List.of("waiting", "received", "unsupported", "unavailable", "error"),
                java.util.Arrays.stream(DetectionStatus.values())
                        .map(value -> mapper.convertValue(value, String.class))
                        .toList());
    }

    @Test
    void exposesTypedSourceCatalogInsteadOfLocalizedTutorialContent() {
        var catalog = new InstrumentationCatalogV2Service(new InstrumentationCatalogService()).catalog();
        JsonNode json = mapper.valueToTree(catalog);

        assertEquals(2, catalog.schemaVersion());
        assertEquals(Set.of("schemaVersion", "groups", "sources", "recipes"), fieldNames(json));
        assertTrue(catalog.sources().size() >= 35);
        assertTrue(catalog.sources().stream()
                .allMatch(source -> source.labelKey().startsWith("instrumentation.v2.directory.")));
        assertTrue(catalog.recipes().stream().allMatch(recipe -> recipe.blocksPreview().size() <= 8));
        assertTrue(catalog.recipes().stream()
                .filter(recipe -> recipe.kind() == SourceKind.APPLICATION)
                .anyMatch(recipe -> "go_ebpf_preview".equals(recipe.id()) && recipe.preview()));
        assertTrue(catalog.recipes().stream()
                .filter(recipe -> recipe.kind() == SourceKind.QUICK_START)
                .flatMap(recipe -> recipe.components().stream())
                .allMatch(component -> component.official() && !component.bundledWithHertzBeat()));
    }

    @Test
    void exposesOnlyExplicitCanonicalHttpsCatalogUrls() {
        var catalog = new InstrumentationCatalogV2Service(new InstrumentationCatalogService()).catalog();

        catalog.recipes().stream().flatMap(recipe -> recipe.components().stream()).forEach(component -> {
            assertExplicitCanonicalHttps(component.sourceUrl());
            component.dependencies().forEach(dependency ->
                    assertExplicitCanonicalHttps(dependency.sourceUrl()));
            component.artifacts().forEach(artifact -> {
                assertExplicitCanonicalHttps(artifact.downloadUrl());
                assertExplicitCanonicalHttps(artifact.provenanceUrl());
            });
        });
    }

    @Test
    void omitsUnsetOptionalRecipeFieldsFromCatalogJson() {
        JsonNode catalog = mapper.valueToTree(
                new InstrumentationCatalogV2Service(new InstrumentationCatalogService()).catalog());
        int recipesWithoutApplicationSelection = 0;

        for (JsonNode recipe : catalog.path("recipes")) {
            if ("quick_start".equals(recipe.path("kind").asText())
                    || "existing_opentelemetry".equals(recipe.path("kind").asText())) {
                recipesWithoutApplicationSelection++;
                assertFalse(recipe.has("language"));
                assertFalse(recipe.has("framework"));
                assertFalse(recipe.has("method"));
            }
        }
        assertTrue(recipesWithoutApplicationSelection >= 2);
    }

    @Test
    void omitsUnsetOptionalDetectionFieldsFromJson() {
        JsonNode waiting = mapper.valueToTree(
                new SignalDetection(DetectionStatus.WAITING, null, DetectionErrorCode.SIGNAL_NOT_RECEIVED));
        JsonNode received = mapper.valueToTree(
                new SignalDetection(DetectionStatus.RECEIVED, 1_710_000_000_000L, null));
        JsonNode complete = mapper.valueToTree(
                new PollingInstruction(PollingDecision.COMPLETE, null, 1_710_000_120_000L));

        assertFalse(waiting.has("lastReceivedAt"));
        assertFalse(received.has("errorCode"));
        assertFalse(complete.has("pollAfterMs"));
    }

    @Test
    void intakeProfileIsIndependentExplicitAndDisclosesTransportSecurity() throws Exception {
        IntakeProfile profile = new IntakeProfile(
                "server-primary",
                IntakeKind.SERVER,
                Availability.AVAILABLE,
                Gateway.SERVER,
                List.of(OtlpTransport.HTTP_PROTOBUF),
                Map.of(
                        OtlpTransport.HTTP_PROTOBUF,
                        new IntakeEndpoint("https://otel.example.test/v1", TransportSecurity.TLS)),
                "Authorization",
                null,
                null);

        String json = mapper.writeValueAsString(profile);
        assertTrue(json.contains("\"id\":\"server-primary\""));
        assertTrue(json.contains("\"kind\":\"server\""));
        assertTrue(json.contains("\"authentication\":\"bearer_token\""));
        assertTrue(json.contains("\"authorizationHeader\":\"Authorization\""));
        assertTrue(json.contains("\"endpoints\":{\"http_protobuf\":"
                + "{\"url\":\"https://otel.example.test/v1\",\"security\":\"tls\"}}"));
        assertFalse(json.contains("httpsEndpoints"));
        assertFalse(json.contains("\"token\":"));
        assertFalse(json.contains("authHeaderName"));
        assertFalse(json.contains("collectorId"));
        IntakeProfile plaintext = new IntakeProfile(
                "collector:loopback",
                IntakeKind.HERTZBEAT_COLLECTOR,
                Availability.AVAILABLE,
                Gateway.COLLECTOR,
                List.of(OtlpTransport.HTTP_PROTOBUF),
                Map.of(
                        OtlpTransport.HTTP_PROTOBUF,
                        new IntakeEndpoint("http://127.0.0.1:4318", TransportSecurity.PLAINTEXT)),
                "Authorization",
                "loopback",
                null);
        assertEquals(
                TransportSecurity.PLAINTEXT,
                plaintext.endpoints().get(OtlpTransport.HTTP_PROTOBUF).security());
        IntakeProfile mixed = new IntakeProfile(
                "collector:mixed",
                IntakeKind.HERTZBEAT_COLLECTOR,
                Availability.AVAILABLE,
                Gateway.COLLECTOR,
                List.of(OtlpTransport.HTTP_PROTOBUF, OtlpTransport.GRPC),
                Map.of(
                        OtlpTransport.HTTP_PROTOBUF,
                        new IntakeEndpoint("http://127.0.0.1:4318", TransportSecurity.PLAINTEXT),
                        OtlpTransport.GRPC,
                        new IntakeEndpoint("https://collector.example.test:4317", TransportSecurity.TLS)),
                "Authorization",
                "mixed",
                null);
        assertEquals(TransportSecurity.PLAINTEXT,
                mixed.endpoints().get(OtlpTransport.HTTP_PROTOBUF).security());
        assertEquals(TransportSecurity.TLS, mixed.endpoints().get(OtlpTransport.GRPC).security());
        assertThrows(IllegalArgumentException.class, () -> new IntakeProfile(
                "unsafe",
                IntakeKind.SERVER,
                Availability.AVAILABLE,
                Gateway.SERVER,
                List.of(OtlpTransport.GRPC),
                Map.of(
                        OtlpTransport.GRPC,
                        new IntakeEndpoint("ftp://explicit-but-unsupported:4317", TransportSecurity.PLAINTEXT)),
                "Authorization",
                null,
                null));
        assertThrows(IllegalArgumentException.class, () ->
                new IntakeEndpoint("http://127.0.0.1:4318", TransportSecurity.TLS));
        assertThrows(IllegalArgumentException.class, () ->
                new IntakeEndpoint("https://otel.example.test:4318", TransportSecurity.PLAINTEXT));
        assertThrows(IllegalArgumentException.class, () -> new IntakeProfile(
                "duplicate",
                IntakeKind.EXTERNAL_OTEL_COLLECTOR,
                Availability.AVAILABLE,
                Gateway.EXTERNAL,
                List.of(OtlpTransport.GRPC, OtlpTransport.GRPC),
                Map.of(
                        OtlpTransport.GRPC,
                        new IntakeEndpoint("https://otel.example.test:4317", TransportSecurity.TLS)),
                "Authorization",
                null,
                null));
        assertThrows(IllegalArgumentException.class, () -> new IntakeProfile(
                "server-none",
                IntakeKind.SERVER,
                Availability.AVAILABLE,
                Gateway.SERVER,
                List.of(OtlpTransport.HTTP_PROTOBUF),
                Map.of(
                        OtlpTransport.HTTP_PROTOBUF,
                        new IntakeEndpoint("https://otel.example.test:4318", TransportSecurity.TLS)),
                Authentication.NONE,
                null,
                null,
                null));
        assertThrows(IllegalArgumentException.class, () -> new IntakeProfile(
                "external-inconsistent",
                IntakeKind.EXTERNAL_OTEL_COLLECTOR,
                Availability.AVAILABLE,
                Gateway.EXTERNAL,
                List.of(OtlpTransport.HTTP_PROTOBUF),
                Map.of(
                        OtlpTransport.HTTP_PROTOBUF,
                        new IntakeEndpoint("https://otel.example.test:4318", TransportSecurity.TLS)),
                Authentication.NONE,
                "Authorization",
                null,
                null));
    }

    private static void assertExplicitCanonicalHttps(String value) {
        URI uri = URI.create(value);
        assertEquals("https", uri.getScheme(), value);
        assertTrue(uri.getHost() != null && !uri.getHost().isBlank(), value);
        assertNull(uri.getUserInfo(), value);
        assertNull(uri.getRawQuery(), value);
        assertNull(uri.getRawFragment(), value);
    }

    private Set<String> fieldNames(JsonNode node) {
        Set<String> names = new HashSet<>();
        node.fieldNames().forEachRemaining(names::add);
        return names;
    }

    @Test
    void secretPlaceholderDeclaresMarkerWithoutCarryingSecretMaterial() throws Exception {
        SecretPlaceholder placeholder = SecretPlaceholder.authorizationToken();

        assertEquals("${HERTZBEAT_TOKEN}", placeholder.marker());
        assertEquals("authorization_token", placeholder.kind());
        assertNull(placeholder.value());
        assertFalse(mapper.writeValueAsString(placeholder).contains("secret-value"));
    }
}
